cloud/linode/cilium_loadbalancers.go

```diff
@@ -0,0 +1,247 @@
+	ipv4PublicAddrs := []string{}
+	addrResp, err := l.client.GetInstanceIPAddresses(ctx, ipHolder.ID)
+	if err != nil {
+		return "", err
+	}
+	for _, addr := range addrResp.IPv4.Public {
+		ipv4PublicAddrs = append(ipv4PublicAddrs, addr.Address)
+	}
+
+	// share the IPs with nodes participating in Cilium BGP peering
+	kv := strings.Split(Options.BGPNodeSelector, "=")
+	for _, node := range nodes {
+		if val, ok := node.Labels[kv[0]]; ok && len(kv) == 2 && val == kv[1] {
+			nodeLinodeID, err := parseProviderID(node.Spec.ProviderID)
+			if err != nil {
+				return "", err
+			}
+
+			if err = l.client.ShareIPAddresses(ctx, linodego.IPAddressesShareOptions{
+				IPs:      ipv4PublicAddrs,
+				LinodeID: nodeLinodeID,
+			}); err != nil {
+				return "", err
+			}
+			klog.Infof("shared IPs %v on Linode %d", ipv4PublicAddrs, nodeLinodeID)
+		}
+	}
+
+	return newSharedIP.Address, nil
+}
+
+// deleteSharedIP cleans up the shared IP for a LoadBalancer Service if it was assigned
+// by Cilium LB IPAM, removing it from the ip-holder
+func (l *loadbalancers) deleteSharedIP(ctx context.Context, service *v1.Service) error {
+	if Options.BGPNodeSelector == "" {
+		return errors.New("no BGP node label set to configure IP sharing")
+	}
+	err := l.retrieveKubeClient()
+	if err != nil {
+		return err
+	}
+	nodeList, err := l.kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+		LabelSelector: Options.BGPNodeSelector,
+	})
+	if err != nil {
+		return err
+	}
+	bgpNodes := nodeList.Items
+	ipHolder, err := l.getIPHolder(ctx)
+	if err != nil {
+		// return error or nil if not found since no IP holder means there
+		// is no IP to reclaim
+		return IgnoreLinodeAPIError(err, http.StatusNotFound)
+	}
+	svcIngress := service.Status.LoadBalancer.Ingress
+	if len(svcIngress) > 0 && ipHolder != nil {
+		for _, ingress := range svcIngress {
+			// delete the shared IP on the Linodes it's shared on
+			for _, node := range bgpNodes {
+				nodeLinodeID, err := parseProviderID(node.Spec.ProviderID)
+				if err != nil {
+					return err
+				}
+				err = l.client.DeleteInstanceIPAddress(ctx, nodeLinodeID, ingress.IP)
+				if IgnoreLinodeAPIError(err, http.StatusNotFound) != nil {
+					return err
+				}
+			}
+
+			// finally delete the shared IP on the ip-holder
+			err = l.client.DeleteInstanceIPAddress(ctx, ipHolder.ID, ingress.IP)
+			if IgnoreLinodeAPIError(err, http.StatusNotFound) != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// To hold the IP in lieu of a proper IP reservation system, a special Nanode is
+// created but not booted and used to hold all shared IPs.
+func (l *loadbalancers) ensureIPHolder(ctx context.Context) (*linodego.Instance, error) {
+	ipHolder, err := l.getIPHolder(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if ipHolder != nil {
+		return ipHolder, nil
+	}
+
+	ipHolder, err = l.client.CreateInstance(ctx, linodego.InstanceCreateOptions{
+		Region:   l.zone,
+		Type:     "g6-nanode-1",
```

Comment:
nit: can we have this configurable via a variable? Though plans don't change frequently, just wondering if it does get updated in future to g7, we can just set the variable than changing and building again. Same for image.

Reply:
This is a temporary workaround while there's no way to request IPs that are not attached to a Linode. I'm hoping that changes by the time there is a g7 plan or ubuntu 22.04 goes EOL 😳